### PR TITLE
add a filter_empty optional param to Dream.query

### PR DESCRIPTION
**Pull request by tatchi**
My attempt at implementing #204

@aantron on a side note, I'm using vscode together with `ocaml-lsp-server` and the official OCaml extension. I wanted to format I've edited but it leads to lots of changes 😅 Do you manually format the code?

### CODE REVIEW

File: src/server/query.ml [hunk 1]
```diff
@@ -28,8 +28,11 @@ let all_queries request =
   |> snd
   |> Formats.from_form_urlencoded
 
-let query request name =
-  List.assoc_opt name (all_queries request)
+let query request name ?(filter_query = false) () =
+  match (List.assoc_opt name (all_queries request), filter_query) with
+  | None, _ -> None
+  | Some v, false -> Some v
+  | Some v, true -> if String.trim v = "" then None else Some v
```
Comment (tatchi) [inline src/server/query.ml:35]: It will `trim` multiple occurrences of spaces. That's not what was described in the original issue but if we only check for one empty space, `" "` will give `None` while multiple spaces will result in `Some("  ")`. It looks a bit weird to me but I'm definitely open to changing it if you prefer.

File: test/expect/pure/formats/query/query.ml [hunk 2]
```diff
@@ -5,22 +5,30 @@
...
+  query ~filter_query:false "c" "a=b&c=d";
+  query ~filter_query:false "a b" "a+b=c";
+  query ~filter_query:false "a" "a";
+  query ~filter_query:true "a" "a";
+  query ~filter_query:true "a" "a ";
+  query ~filter_query:true "a" "a  ";
+  query ~filter_query:true "a" "a=";
+  query ~filter_query:true "a" "a= ";
+  query ~filter_query:true "a" "a=  ";
+  query ~filter_query:true "a" "a= &a=b";
```
Comment (tatchi) [inline test/expect/pure/formats/query/query.ml:31]: That one currently gives `None`. Is that what we want?

File: src/dream.mli [hunk 3]
```diff
@@ -414,7 +414,7 @@ https://github.com/aantron/dream/issues
 (** Replaces the path. See {!Dream.val-path}. *)
 (**/**)
 
-val query : request -> string -> string option
+val query : request -> string -> ?filter_query:bool -> unit -> string option
```
Comment (ulrikstrid) [inline src/dream.mli:417]: If you move the optional parameter to the left you can skip the final unit argument. Then you don't have to change the examples. 🙂

Comment (tatchi) [reply src/dream.mli]: Beautiful, thanks! 🤩

File: src/dream.mli [hunk 4]
```diff
@@ -414,7 +414,7 @@ https://github.com/aantron/dream/issues
 (** Replaces the path. See {!Dream.val-path}. *)
 (**/**)
 
-val query : request -> string -> string option
+val query : request -> ?filter_empty:bool -> string -> string option
```
Comment (aantron) [inline src/dream.mli:417]: Could you move the optional fully to the left?

File: src/server/query.ml [hunk 5]
```diff
@@ -28,8 +28,11 @@ let all_queries request =
   |> snd
   |> Formats.from_form_urlencoded
 
-let query request name =
-  List.assoc_opt name (all_queries request)
+let query request ?(filter_empty = false) name =
+  match (List.assoc_opt name (all_queries request), filter_empty) with
```
Comment (aantron) [inline src/server/query.ml:32]: I recommend rewriting this as

```ocaml
match List.assoc_opt name (all_queries request) with
| None -> None
| Some v as result ->
  if filter_empty then
    result
  else
    if String.trim v = "" then
      None
    else
      result
```

This avoids unconditionally allocating `Some` options every time there is a `Some v` match. Also, it avoids mixing in the `filter_empty` check into all branches, when it is only relevant for the `Some v` branch.

Later, if actually optimizing, we may want to replace `String.trim` by a scan of the string that does not allocate an intermediate string potentially almost as long as `v`, only for us to compare it with `""`.

File: test/expect/pure/formats/query/query.ml [hunk 6]
```diff
@@ -5,8 +5,8 @@
 
 
 
-let query name string =
-  Dream.query (Dream.request ~target:("/?" ^ string) "") name
+let query name ?(filter_empty) string =
```
Comment (aantron) [inline test/expect/pure/formats/query/query.ml:8]: The parentheses are not necessary here.

File: test/expect/pure/formats/query/query.ml [hunk 7]
```diff
@@ -21,6 +21,14 @@ let%expect_test _ =
...
   query "c" "a=b&c=d";
   query "a b" "a+b=c";
+  query "a" "a";
+  query ~filter_empty:true "a" "a";
+  query ~filter_empty:true "a" "a ";
+  query ~filter_empty:true "a" "a  ";
+  query ~filter_empty:true "a" "a=";
+  query ~filter_empty:true "a" "a= ";
+  query ~filter_empty:true "a" "a=  ";
+  query ~filter_empty:true "a" "a= &a=b";
```
Comment (aantron) [inline test/expect/pure/formats/query/query.ml:31]: Can you add `query ~filter_empty:true "a" "a= b "` and any other tests that use `~filter_empty` but should *not* produce `None`?

**Review by aantron — COMMENTED** [review]
Thank you! Just have a few notes and nits.

**Comment by aantron** [comment]
(But note, if you'd like to format some code you're adding, Dream does have an [`.ocamlformat`](https://github.com/aantron/dream/blob/master/.ocamlformat) in its repo that produces some approximation of a modern, diff-friendly, and Web-oriented style, to the extent that's currently possible with Ocamlformat.)

File: test/expect/pure/formats/query/query.ml (hunk 2 above)
Comment (aantron) [reply test/expect/pure/formats/query/query.ml]: Yes. [`Dream.query`](https://aantron.github.io/dream/#val-query):

> First query parameter with the given name.

[`Dream.queries`](https://aantron.github.io/dream/#val-queries) can be used to get the result `[" ", "b"]` here.

File: src/server/query.ml (hunk 5 above)
Comment (tatchi) [reply src/server/query.ml]: > This avoids unconditionally allocating Some options every time there is a Some v match. Also, it avoids mixing in the filter_empty check into all branches, when it is only relevant for the Some v branch.

Oh, that makes sense, thanks for the explanation 😊 I think the conditions should be inverted in your example. It should be:

```ocaml
match List.assoc_opt name (all_queries request) with
| None -> None
| Some v as result ->
  if filter_empty then
    if String.trim v = "" then
      None
    else
      result
  else
    result
```

Which I've simplified to:

```ocaml
match List.assoc_opt name (all_queries request) with
| None -> None
| Some v as result ->
  if filter_empty && String.trim v = "" then
    None
  else
    result
```

I hope that makes sense.

Comment (tatchi) [reply src/server/query.ml]: > Later, if actually optimizing, we may want to replace String.trim by a scan of the string that does not allocate an intermediate string potentially almost as long as v, only for us to compare it with "".

Interesting, would something like the following be reasonable ?

```ocaml
(* taken from https://github.com/ocaml/ocaml/blob/trunk/stdlib/string.ml#L97 *)
let is_space = function
  | ' ' | '\012' | '\n' | '\r' | '\t' -> true
  | _ -> false
  
let is_empty s =
  let len = String.length s in
  if len = 0 then
    true
  else
    let rec loop len i =
      if i < len && is_space s.[i] then
        loop len (i + 1)
      else
        false
    in
    loop len 0
```

Or without the inner `loop` function:

```ocaml
let is_empty s =
  let len = String.length s in
  if len = 0 then
    true
  else
    let i = ref 0 in
    let contains_only_spaces = ref true in
    while !contains_only_spaces && !i < len do
      if is_space s.[!i] then
        i := !i + 1
      else
        contains_only_spaces := false
    done;
    !contains_only_spaces
```

**Comment by tatchi** [comment]
Nice to see you back working on this project @aantron 🤗😍

**Comment by aantron** [comment]
Thank you :)

**Comment by aantron** [comment]
After the huge delay (apologies!) I think this behavior is too niche to support directly in Dream, given a user can easily write their own wrapper to deal with the inconveniences of their client's behavior. I'm going to close this, but we can resurrect it if we have claims otherwise. Sorry and thank you!
